### DISCUSSION TITLE
Which audio settings would keep sound contained in one spot and not spill out into the space?

### DISCUSSION THREAD

**Discussion by hilarytsai** [mozilla/hubs#3173] [Spoke Questions] [answered]
I know this was reported as a bug some time ago, and I've heard it's since been fixed.  But I'm having some trouble with sound, still.  My space has music that's supposed to sound that's like it's coming from one area, so it should taper off as you move away from it.  I've played with the various settings on the file like rolloff, reference distance, cone angle and it seems like the sound is audible in the rest of the space no matter what I try.  Part of it is that I have no real frame reference, or what 1 meter represents in my space, and have trouble wrapping my head around these settings.  For example, rolloff=the higher the value, the more quickly it fades?  Should reference distance generally be higher or lower than the 1m default?  Should max distance most likely be below or above the default 100m?  And which distance model should I use for this situation?  

I feel like I've tried lots of different settings, and have been publishing each time to see changes, which takes a long time.  I can't remember whether the preview feature in Spoke reflects audio settings, but it doesn't seem to in my experience.  If it's ok, I will include my file if anyone is able to take a look at it.  The audio file in question is at the very bottom of the list, an mp3 called "Jonas Franklin."  It's right in front of the book store, and I've kept it visible so you can see.  I'd really appreciate if anyone can help me out with the settings, as I've been stuck on it for quite a while.  The music should taper off pretty quickly as you move away from the book store front. 

https://hubs.mozilla.com/scenes/SWbWrse

Thank you so much!

**Comment by vincentfretin** [comment]
Hi, I don't know much about Hubs and the various sounds parameters myself, I'm just pointing to links I came across.
Relevant screen in Spoke : https://github.com/mozilla/hubs/discussions/3153#discussioncomment-95575
And you may find this article useful https://medium.com/@kfarr/understanding-web-audio-api-positional-audio-distance-models-for-webxr-e77998afcdff

**Comment by johnshaughnessy** [comment, marked as answer]
@hilarytsai These are great questions. We know we need to make these settings more obvious and easier to understand/test in Spoke, but have not quite gotten there yet. For now though, I'll share a few recommendations for the specific thing you're trying to accomplish.

##### Background Info:
The settings we expose in Spoke are for configuring things called `PannerNode`s which are implemented by browsers to give websites the ability to control audio experiences. This may be more technical/detailed than what you're looking for, but if you are curious about exactly what each of the parameters mean, [this documentation](https://developer.mozilla.org/en-US/docs/Web/API/PannerNode) goes into that detail.  

##### Note on Overrides: 
On the root node of your scene, make sure "Override Audio Settings" is **not** checked. If it is checked, any settings you put on specific nodes will be overridden. (This is not a problem in the scene you shared, but in case other people read this discussion, I included it here.)
![image](https://user-images.githubusercontent.com/4072106/96181493-f032e180-0ee8-11eb-9085-12fd241775ab.png)

##### Note on Rolloff Factor:
We discovered a bug last week (that needs to be investigated and filed in an issue) where the `Rolloff Factor` was not working correctly in Chrome with the `linear` distance model. For this reason, I recommend setting the `Rolloff Factor` to `1` (the default setting).

##### Note on visibility:
We have[ an update going out this week](https://github.com/mozilla/hubs/pull/3121) to fix a bug where invisible audio nodes end up at the origin of the space (at the point `0,0,0` instead of the `x,y,z` that you configure). This probably contributed to your difficulties in finding the right settings. Until this bug is fixed, you probably want to debug audio with the visibility turned on:
![image](https://user-images.githubusercontent.com/4072106/96183193-66d0de80-0eeb-11eb-9a59-7bdac6c1cb68.png)

##### Specific Recommendation:
> Which audio settings would keep sound contained in one spot and not spill out into the space?
> Which distance model should I use for this situation?

The settings that I find easiest to configure and understand are:

| Setting | Value | Reason |
| -----------: | :--------- | :--------- |
| Distance Model | linear | Easy to understand because you just configure its min/max volume and where it starts and where it ends. |
| Rolloff Factor | 1 | This means it will go from max volume to min volume across the distance you specify |
| Ref Distance | 1 | This means that the volume will be at 100% within a meter of the source object |
| Max Distance | 10 | This means that the volume will be at 0% when you are 10 meters away from the source object |
| Cone Inner Angle | 360 | This means the audio spreads out in all directions (I think) |
| Cone Outer Angle | 360 | Same as above |
| Cone Outer Gain | 0 (or anything) | This only takes effect if the inner/outer angles are set differently. See docs linked above for details. |

There may be better settings that you'll find by tinkering with either the distance model or the specific distances. I suspect your difficulties are largely due to the "invisible audio nodes" bug I mentioned above, and possibly also the `rolloff factor` not working as intended in Chrome.

**Reply by johnshaughnessy** [reply]
> I feel like I've tried lots of different settings, and have been publishing each time to see changes, which takes a long time.

This is a great point. I'd like to add a way to test these audio settings more rapidly in the client. Most of the code we'd need is already there -- we'd just need a UI for doing it and clear instructions for how to get those settings back into Spoke. 
Or maybe, we need to look at the diff between the previously exported project and the newly exported project and only require your Spoke client to upload the changes (which would happen instantly if you were only changing audio settings). This approach is probably more difficult but would likely be a better experience overall.
This is worth making a feature request in either the Spoke repo or the Hubs repo because I agree it would be a big quality of life improvement.

**Reply by hilarytsai** [reply]
Thank you so much @johnshaughnessy!  This really breaks down what i need to know about the audio settings and makes them easier to understand.  I will be tweaking the sound with all of this in mind, and I think my teammates for this project will appreciate the info as well.   It's also great to hear that a way to test sound within the client could be developed--I'll probably put in a feature request about it, as I agree it would make things much more efficient.  Thanks again!  Your explanations are helpful and appreciated:)

**Comment by j-conrad** [comment]
Adding some new info to this discussion because there have been some advancements in audio since this thread started:

_This PR adds support for Audio Zones, a 3D box volume that modifies the audio properties of audio sources (avatars, videos and audios) based on the source's and listener's positions with respect to the audio zone. One obvious application would be to dim audio sources' volumes based on 3D areas like rooms to mimic the real world behavior._

[Audio Zones](https://github.com/mozilla/hubs/pull/4399)
